token/transfer-hook-interface/src/instruction.rs

```diff
@@ -0,0 +1,206 @@
+//! Instruction types
+
+use {
+    solana_program::{
+        instruction::{AccountMeta, Instruction},
+        program_error::ProgramError,
+        pubkey::Pubkey,
+        system_program,
+    },
+    spl_type_length_value::discriminator::{Discriminator, TlvDiscriminator},
+    std::convert::TryInto,
+};
+
+/// Instructions supported by the transfer hook interface.
+#[repr(C)]
+#[derive(Clone, Debug, PartialEq)]
+pub enum TransferHookInstruction {
+    /// Runs additional transfer logic.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   0. `[]` Source account
+    ///   1. `[]` Token mint
+    ///   2. `[]` Destination account
+    ///   3. `[]` Source account's owner/delegate
+    ///   4. `[]` Validation account
+    ///   5..5+M `[]` `M` additional accounts, written in validation account data
+    ///
+    Execute {
+        /// Amount of tokens to transfer
+        amount: u64,
+    },
```

Comment:
Having an interface that only gets access to token accounts seems weird, since we may want to ~steal~ charge royalties in sol against the owner wallet in the hook program. That feels like a pretty natural usage

Reply:
The interface states it gets the transfer accounts PLUS all of the additional accounts specified, so it doesn't just get the transfer accounts. That's what the `    ///   5..5+M `[]` `M` additional accounts, written in validation account data` line is about. Maybe it needs to be clearer?

Reply:
This might not be related but the way you've written the comments reminds me of the nomenclature for passing multisig accounts to a token instruction, ie:
```rust
    /// Accounts expected by this instruction:
    ///
    ///   * Single owner/delegate
    ///   0. `[writable]` The source account.
    ///   1. `[writable]` The destination account.
    ///   2. `[signer]` The source account's owner/delegate.
    ///
    ///   * Multisignature owner/delegate
    ///   0. `[writable]` The source account.
    ///   1. `[writable]` The destination account.
    ///   2. `[]` The source account's multisignature owner/delegate.
    ///   3. ..3+M `[signer]` M signer accounts.
    Transfer {
        /// The amount of tokens to transfer.
        amount: u64,
    },
```

Have we considered how both of these variably-sized additional accounts might work together?

Reply:
Starting to think they would just be included, but would have to be signers, no?

Reply:
The idea is to keep it simple for multisigs, and just provide the authority *without* signers. The token program won't provide any existing of the token accounts as signer or writable to be safe